src/Simulation/QCTraceSimulator/SortedQubitPool.cs

```diff
@@ -0,0 +1,329 @@
+            }
+        }
+
+        /// <summary>
+        /// Compares two QubitTimeNode objects for SortedSet.
+        /// It also collects minimum and maximum Time value of nodes been compared to the
+        /// sample Time since ResetForComparison is called.
+        /// </summary>
+        private class VisitingComparer : IComparer<QubitTimeNode>
+        {
+            /// <summary>
+            /// Maximum of all values, that are less than or equal than sample, among values seen after reset.
+            /// </summary>
+            internal ComplexTime MaxLowerBound = ComplexTime.MinValue;
+
+            /// <summary>
+            /// Minimum of all values, that are greater or equal than sample, among values seen after reset.
+            /// </summary>
+            internal ComplexTime MinUpperBound = ComplexTime.MaxValue;
+
+            /// <summary>
+            /// Sample, that will be passed to a series of comparisons.
+            /// Sample is not an object that's stored in the set so it is excluded from the values seen.
+            /// </summary>
+            private QubitTimeNode Sample;
+
+            /// <summary>
+            /// Resets minimum and maximum value of nodes seen from the set.
+            /// Also sets sample object, which should not be counted because it is not an object from the set.
+            /// </summary>
+            /// <param name="sample">Sample object that is not considered for minimum and maximum.</param>
+            internal void ResetForComparison(QubitTimeNode sample) {
+                MaxLowerBound = ComplexTime.MinValue;
+                MinUpperBound = ComplexTime.MaxValue;
+                Sample = sample;
+            }
+
+            /// <summary>
+            /// Compares Time field of two nodes. Also updates MaxLowerBound and MinUpperBound.
+            /// </summary>
+            /// <returns>Result of comparison of Time field of QubitTimeNode</returns>
+            public int Compare(QubitTimeNode a, QubitTimeNode b) {
+                // Note that comparison of a and b nodes to Sample is "by reference" in this function.
+                int result = ComplexTime.Compare(a.Time, b.Time);
+                if (result > 0) {
+                    if (a == Sample && b != Sample) {
+                        MaxLowerBound = ComplexTime.Max(MaxLowerBound, b.Time);
+                    } else if (a != Sample && b == Sample) {
+                        MinUpperBound = ComplexTime.Min(MinUpperBound, a.Time);
+                    }
+                } else if (result < 0) {
+                    if (a == Sample && b != Sample) {
+                        MinUpperBound = ComplexTime.Min(MinUpperBound, b.Time);
+                    } else if (a != Sample && b == Sample) {
+                        MaxLowerBound = ComplexTime.Max(MaxLowerBound, a.Time);
+                    }
+                } else {
+                    // We found the value if one argument is the sample and nother is not.
```

Comment:
typo: nother -> other

Reply:
also, why cannot both `a` and `b` be equal to Sample?

Reply:
Updated comment.

Reply:
It is possible that both a and b would be equal to sample by reference. This would be a useless comparison. And it is handled correctly. This doesn't give us any information about members of the set so such comparison is ignored (and correct result is returned).